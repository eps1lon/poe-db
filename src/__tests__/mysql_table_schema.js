const MysqlTableSchema = require('../MysqlTableSchema');

const spec = require('./spec.json')[0];

describe('MysqlTableSchema', () => {
  const mods = new MysqlTableSchema('Mods.dat', spec['Mods.dat']);

  it('should handle null fields', () => {
    expect(
      new MysqlTableSchema(
        'OldMapStashTabLayout.dat',
        spec['OldMapStashTabLayout.dat'],
      ).fields,
    ).toEqual([]);
  });

  it('should recognized extended props', () => {
    expect(mods.isExtendedProp('SpawnWeight_TagsKeys')).toBe(true);
    expect(mods.isExtendedProp('SpawnWeight_Values')).toBe(true);
    expect(mods.isExtendedProp('SomeExtendedProp_TagsKey')).toBe(false);
    expect(mods.isExtendedProp('SomeSimpleProp')).toBe(false);
  });

  it('should recognize hasMany relations', () => {
    expect(mods.isHasMany('SpawnWeight_TagsKeys')).toBe(true);
    expect(mods.isHasMany('Keys0')).toBe(true);
    expect(mods.isHasMany('Key0')).toBe(false);
  });

  it('should recognize foreign keys', () => {
    expect(mods.isForeignKey('SpawnWeight_TagsKeys')).toBe(true);
    expect(mods.isForeignKey('BuffDefinitionsKey')).toBe(true);
    expect(mods.isForeignKey('Keys0')).toBe(true);
    expect(mods.isForeignKey('Key0')).toBe(true);
  });

  it('should recognize key candidates', () => {
    expect(mods.isKeyCandidate('Row')).toBe(true);
    expect(mods.isKeyCandidate('SpawnWeight_TagsKeys')).toBe(true);
    expect(mods.isKeyCandidate('BuffDefinitionsKey')).toBe(true);
    expect(mods.isKeyCandidate('Keys0')).toBe(true);
    expect(mods.isKeyCandidate('Key0')).toBe(true);
  });

  it('should output correct col definitions', () => {
    expect(mods.columnDefinition('Row')).toBe(
      'BIGINT UNSIGNED NOT NULL PRIMARY KEY',
    );
    expect(mods.columnDefinition('SpawnWeight_TagsKeys')).toBe(
      'BIGINT UNSIGNED NOT NULL PRIMARY KEY',
    );
    expect(mods.columnDefinition('BuffDefinitionsKey')).toBe(
      'BIGINT UNSIGNED NOT NULL',
    );
    expect(mods.columnDefinition('Id')).toBe('TEXT NOT NULL');
    expect(mods.columnDefinition('Stat1Min')).toBe('INT NOT NULL');
  });

  it('should generate correct col names', () => {
    expect(mods.colName('Row')).toBe('`row`');
    expect(mods.colName('SpawnWeight_TagsKeys')).toBe('`tags_row`');
    expect(mods.colName('BuffDefinitionsKey')).toBe('`buff_definitions_row`');
    expect(mods.colName('Id')).toBe('`id`');
    expect(mods.colName('Stat1Min')).toBe('`stat1_min`');
  });

  it('should generate correct create queries', () => {
    expect(mods.createQuery()).toMatchSnapshot();

    expect(
      new MysqlTableSchema(
        'MonsterVarieties.dat',
        spec['MonsterVarieties.dat'],
      ).createQuery(),
    ).toMatchSnapshot();
  });

  it('should generate correct create queries for all relations', () => {
    expect(mods.createQueryRelations()).toMatchSnapshot();

    expect(
      new MysqlTableSchema(
        'MonsterVarieties.dat',
        spec['MonsterVarieties.dat'],
      ).createQueryRelations(),
    ).toMatchSnapshot();

    expect(
      new MysqlTableSchema(
        'AtlasNode.dat',
        spec['AtlasNode.dat'],
      ).createQueryRelations(),
    ).toMatchSnapshot();
  });
});
