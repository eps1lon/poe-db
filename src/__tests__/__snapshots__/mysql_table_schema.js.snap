// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MysqlTableSchema should generate correct create queries 1`] = `
"CREATE TABLE \`mods\` (row BIGINT UNSIGNED NOT NULL PRIMARY KEY,
id TEXT NOT NULL,
unknown0 INT NOT NULL,
mod_type_row BIGINT UNSIGNED NOT NULL,
INDEX (mod_type_row),
level INT NOT NULL,
stats1_row BIGINT UNSIGNED NOT NULL,
INDEX (stats1_row),
stats2_row BIGINT UNSIGNED NOT NULL,
INDEX (stats2_row),
stats3_row BIGINT UNSIGNED NOT NULL,
INDEX (stats3_row),
stats4_row BIGINT UNSIGNED NOT NULL,
INDEX (stats4_row),
domain INT NOT NULL,
name TEXT NOT NULL,
generation_type INT NOT NULL,
correct_group TEXT NOT NULL,
stat1_min INT NOT NULL,
stat1_max INT NOT NULL,
stat2_min INT NOT NULL,
stat2_max INT NOT NULL,
stat3_min INT NOT NULL,
stat3_max INT NOT NULL,
stat4_min INT NOT NULL,
stat4_max INT NOT NULL,
buff_definitions_row BIGINT UNSIGNED NOT NULL,
INDEX (buff_definitions_row),
buff_value INT NOT NULL,
granted_effects_per_level_row BIGINT UNSIGNED NOT NULL,
INDEX (granted_effects_per_level_row),
data1 TEXT NOT NULL,
data2 TEXT NOT NULL,
monster_metadata TEXT NOT NULL,
data3 TEXT NOT NULL,
data4 TEXT NOT NULL,
0_row BIGINT UNSIGNED NOT NULL,
INDEX (0_row),
stat5_min INT NOT NULL,
stat5_max INT NOT NULL,
stats5_row BIGINT UNSIGNED NOT NULL,
INDEX (stats5_row),
achievement_items_row BIGINT UNSIGNED NOT NULL,
INDEX (achievement_items_row),
data5 TEXT NOT NULL,
is_essence_only_modifier BOOLEAN NOT NULL);"
`;

exports[`MysqlTableSchema should generate correct create queries 2`] = `
"CREATE TABLE \`monster_varieties\` (row BIGINT UNSIGNED NOT NULL PRIMARY KEY,
id TEXT NOT NULL,
monster_types_row BIGINT UNSIGNED NOT NULL,
INDEX (monster_types_row),
unknown0 INT NOT NULL,
object_size INT NOT NULL,
minimum_attack_distance INT NOT NULL,
maximum_attack_distance INT NOT NULL,
act_file TEXT NOT NULL,
ao_file TEXT NOT NULL,
base_monster_type_index TEXT NOT NULL,
unknown1 INT NOT NULL,
unknown_index0 TEXT NOT NULL,
unknown_index1 TEXT NOT NULL,
model_size_multiplier INT NOT NULL,
unknown2 INT NOT NULL,
unknown3 INT NOT NULL,
unknown4 INT NOT NULL,
unknown5 INT NOT NULL,
unknown6 INT NOT NULL,
experience_multiplier INT NOT NULL,
unknown7 TEXT NOT NULL,
unknown8 INT NOT NULL,
unknown9 INT NOT NULL,
unknown10 INT NOT NULL,
critical_strike_chance INT NOT NULL,
unknown11 INT NOT NULL,
ais_file TEXT NOT NULL,
stance TEXT NOT NULL,
2_row BIGINT UNSIGNED NOT NULL,
INDEX (2_row),
name TEXT NOT NULL,
damage_multiplier INT NOT NULL,
life_multiplier INT NOT NULL,
attack_speed INT NOT NULL,
back_item_visual_identity_row BIGINT UNSIGNED NOT NULL,
INDEX (back_item_visual_identity_row),
main_hand_item_classes_row BIGINT UNSIGNED NOT NULL,
INDEX (main_hand_item_classes_row),
off_hand_item_classes_row BIGINT UNSIGNED NOT NULL,
INDEX (off_hand_item_classes_row),
1_row BIGINT UNSIGNED NOT NULL,
INDEX (1_row),
helmet_item_visual_identity_row BIGINT UNSIGNED NOT NULL,
INDEX (helmet_item_visual_identity_row),
unknown13 INT NOT NULL,
unknown14 BOOLEAN NOT NULL,
unknown15 INT NOT NULL,
unknown16 INT NOT NULL,
unknown17 INT NOT NULL,
unknown18 INT NOT NULL,
unknown19 INT NOT NULL,
unknown20 INT NOT NULL,
unknown21 INT NOT NULL,
unknown22 BIT(8) NOT NULL,
unknown23 TEXT NOT NULL,
flashback_achievement_items_row BIGINT UNSIGNED NOT NULL,
INDEX (flashback_achievement_items_row),
monster_segments_row BIGINT UNSIGNED NOT NULL,
INDEX (monster_segments_row),
monster_armours_row BIGINT UNSIGNED NOT NULL,
INDEX (monster_armours_row),
achievement_items_row BIGINT UNSIGNED NOT NULL,
INDEX (achievement_items_row),
achievement_items4_row BIGINT UNSIGNED NOT NULL,
INDEX (achievement_items4_row),
3_row BIGINT UNSIGNED NOT NULL,
INDEX (3_row));"
`;

exports[`MysqlTableSchema should generate correct create queries for all relations 1`] = `
"CREATE TABLE \`mods_spawn_weight_tags\` (mods_row BIGINT UNSIGNED,
tags_row BIGINT UNSIGNED,
values INT UNSIGNED);

CREATE TABLE \`mods_tags\` (mods_row BIGINT UNSIGNED,
tags_row BIGINT UNSIGNED);

CREATE TABLE \`mods0\` (mods_row BIGINT UNSIGNED,
row BIGINT UNSIGNED);

CREATE TABLE \`mods_generation_weight_tags\` (mods_row BIGINT UNSIGNED,
tags_row BIGINT UNSIGNED,
values INT);"
`;

exports[`MysqlTableSchema should generate correct create queries for all relations 2`] = `
"CREATE TABLE \`monster_varieties_mods\` (monster_varieties_row BIGINT UNSIGNED,
mods_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_tags\` (monster_varieties_row BIGINT UNSIGNED,
tags_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_granted_effects\` (monster_varieties_row BIGINT UNSIGNED,
granted_effects_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_mods2\` (monster_varieties_row BIGINT UNSIGNED,
mods_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_weapon1_item_visual_identity\` (monster_varieties_row BIGINT UNSIGNED,
item_visual_identity_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_weapon2_item_visual_identity\` (monster_varieties_row BIGINT UNSIGNED,
item_visual_identity_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_achievement_items\` (monster_varieties_row BIGINT UNSIGNED,
achievement_items_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_special_mods\` (monster_varieties_row BIGINT UNSIGNED,
mods_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_achievement_items2\` (monster_varieties_row BIGINT UNSIGNED,
achievement_items_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_normal_achievement_items\` (monster_varieties_row BIGINT UNSIGNED,
achievement_items_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_merciless_achievement_items\` (monster_varieties_row BIGINT UNSIGNED,
achievement_items_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_cruel_achievement_items\` (monster_varieties_row BIGINT UNSIGNED,
achievement_items_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_normal_mods\` (monster_varieties_row BIGINT UNSIGNED,
mods_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_cruel_mods\` (monster_varieties_row BIGINT UNSIGNED,
mods_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_merciless_mods\` (monster_varieties_row BIGINT UNSIGNED,
mods_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_kill_rare_in_normal_achievement_items\` (monster_varieties_row BIGINT UNSIGNED,
achievement_items_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_kill_rare_in_cruel_achievement_items\` (monster_varieties_row BIGINT UNSIGNED,
achievement_items_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_kill_rare_in_merciless_achievement_items\` (monster_varieties_row BIGINT UNSIGNED,
achievement_items_row BIGINT UNSIGNED);

CREATE TABLE \`monster_varieties_count_kills_achievement_items\` (monster_varieties_row BIGINT UNSIGNED,
achievement_items_row BIGINT UNSIGNED);"
`;
